### ISSUE TITLE
Enum assignability false positive

### ISSUE THREAD

**Issue by UselessPickles** [microsoft/TypeScript#29879]
**TypeScript Version:**  3.3.3

**Search Terms:** enum assignable assignability

**Code**

```ts
namespace Foo {
    export enum Color {
        RED = "RED",
        GREEN = "GREEN"
    }
}

namespace Bar {
    export enum Color {
        RED = "RED",
        // value doesn't even match between types
        GREEN = "BLARGH!!!",
        BLUE = "BLUE"
    }
}

declare const fooColor: Foo.Color;
// expected compiler error here because I'm assigning one enum type to another
const barColor: Bar.Color = fooColor;
```

**Expected behavior:**

Type Foo.Color should not be assignable to type Bar.Color, because they are different enums.

**Actual behavior:**

No compiler error.

The pattern seems to be that the enum will be assignable if:
* Both enum types have the same name ("Color" in this example). If I change the name of one of the enum types, then I get a compiler error as expected.
* And the destination type contains a superset (or exact match) of source enum's value names. If source type contains an entry that does not exist in the destination type, then it is correctly unassignable.

Strangely, the values of overlapping enum names don't even have to match between the enum types (Look at `Foo.Color.GREEN` vs `Bar.Color.GREEN`).

**Playground Link:** https://www.typescriptlang.org/play/#src=namespace%20Foo%20%7B%0D%0A%20%20%20%20export%20enum%20Color%20%7B%0D%0A%20%20%20%20%20%20%20%20RED%20%3D%20%22RED%22%2C%0D%0A%20%20%20%20%20%20%20%20GREEN%20%3D%20%22GREEN%22%0D%0A%20%20%20%20%7D%0D%0A%7D%0D%0A%0D%0Anamespace%20Bar%20%7B%0D%0A%20%20%20%20export%20enum%20Color%20%7B%0D%0A%20%20%20%20%20%20%20%20RED%20%3D%20%22RED%22%2C%0D%0A%20%20%20%20%20%20%20%20%2F%2F%20value%20doesn't%20even%20match%20between%20types%0D%0A%20%20%20%20%20%20%20%20GREEN%20%3D%20%22BLARGH!!!%22%2C%0D%0A%20%20%20%20%20%20%20%20BLUE%20%3D%20%22BLUE%22%0D%0A%20%20%20%20%7D%0D%0A%7D%0D%0A%0D%0Adeclare%20const%20fooColor%3A%20Foo.Color%3B%0D%0A%2F%2F%20expected%20compiler%20error%20here%20because%20I'm%20assigning%20one%20enum%20type%20to%20another%0D%0Aconst%20barColor%3A%20Bar.Color%20%3D%20fooColor%3B%0D%0A

**Related Issues:**

**Comment by jaredpalmer** [comment]
Can confirm this.

**Comment by UselessPickles** [comment]
What is the intended rule for assignability between string enum types? 

I would expect it to be based on the "shapes" of the enum definitions matching: if both enum types have the same exact name="value" pairs defined, then values should be assignable between the two enum types.

For example, the framework our API team uses to generate swagger (OpenApi) specifications unfortunately can't describe shared enum definitions used across multiple fields of multiple models. If the API code has one enum defined that is shared across many fields of many models, then the definition will be duplicated for every field of every model in the swagger document, and will be named differently (based on the name of the model and field it is defined for), but have the same exact name="value" pairs.

It would be very obnoxious if I couldn't assign values between all of these identical enum definitions  because they are technically different types and/or because they have different names.

**Comment by vemoo** [comment]
I found some other cases that seem related to this. It also happens if one is an string enum and the other is a number enum:
```ts
namespace Foo {
    export enum Color {
        RED = "RED",
        GREEN = "GREEN",
        BLUE = "BLUE",
    }
}

namespace Bar {
    export enum Color {
        RED = 0,
        GREEN = 1,
        BLUE = 2,
    }
}

{
    const fooColor: Foo.Color = Foo.Color.BLUE;
    // OK: Compile error (it seems to treat it as the Foo.Color.BLUE literal type)
    const barColor: Bar.Color = fooColor;
}

{
    const fooColor: Foo.Color = Foo.Color.BLUE as Foo.Color;
    // ERROR: Compiles but shouldn't (now it cannot treat it as the Foo.Color.BLUE literal type)
    const barColor: Bar.Color = fooColor;
}

interface Colors {
    fooColor: Foo.Color;
    barColor: Bar.Color;
}

const colors: Colors = {
    // ERROR: Compiles but shouldn't
    barColor: Foo.Color.BLUE,
    // ERROR: Compiles but shouldn't
    fooColor: Bar.Color.BLUE,
};
```
[Playground link](http://www.typescriptlang.org/play/?ts=3.7-Beta#code/HYQwtgpgzgDiDGEAEAxA9mpBvAUE-SEAHjGgE4AuhwArmEgMJoA252eBnASgKIAiSALxIARLz4iANB074A4rx4A5IaIU9lUmbIBCAGQCqPVSP1GtnAL45rOUJFgJkOkGXadipStTqMWbXFkCcVUABmkg+UUVYQBGCMizY2EAJgT8a1tAgng0YCgqADMMJlYyAC5UDAA6UrZhdDRa-zJqpIBubQB6LqQAeQBpSqYwGABLZmQIMjI2AAoxqigICDAoJApMCjIIECpFpBB1igALZEbmsrbDY2ZF6ZBmDYBPGAgASm1c-KoAI1c6hUkC5WoDVMU0IDOlkvnkCkgIYDKhcwQ0aoDrkZDusUS1OpwekgeFwuH0uMM0KMJtAkL8aEsTmgaMwACbAADkVDmwDQAHckAd4CBgDyqNtdvsqEcNmcqk0MUkkHcKA8nhRXh9YT9aQCWpUQZd6giSnibDgcGNgCqyIUnH4yutsvhEXq5YayPiCP8yEjga53dDzd94bkHRSHaonUhCcTSeS-FTJus6QymayORRtN7fbirkl0tHerGyRTEzSU0goIzmWzOdoXWV9f6FTdpJZ2kA)

Another weird thing, if the enums have different names it works as expected:
```ts
namespace Foo {
    export enum FooColor {
        RED = "RED",
        GREEN = "GREEN",
        BLUE = "BLUE",
    }
}

namespace Bar {
    export enum BarColor {
        RED = 0,
        GREEN = 1,
        BLUE = 2,
    }
}

{
    const fooColor: Foo.FooColor = Foo.FooColor.BLUE;
    // OK: Compile error
    const barColor: Bar.BarColor = fooColor;
}

{
    const fooColor: Foo.FooColor = Foo.FooColor.BLUE as Foo.FooColor;
    // OK: Compile error
    const barColor: Bar.BarColor = fooColor;
}

interface Colors {
    fooColor: Foo.FooColor;
    barColor: Bar.BarColor;
}

const colors: Colors = {
    // OK: Compile error
    barColor: Foo.FooColor.BLUE,
    // OK: Compile error
    fooColor: Bar.BarColor.BLUE,
};
```
[Playground link](http://www.typescriptlang.org/play/?ts=3.7-Beta&ssl=1&ssc=1&pln=39&pc=3#code/HYQwtgpgzgDiDGEAEAxA9mpBvAUE-SEAHjGgE4AuhwArmKhgMJoA252eBXASgKIAiSALxIARH36iANJy74A4n14A5YWMW8V02XIBCAGQCqvNaIPHtXAL44bOUJFgJkukGQ5dipStTpJXZMxs7rhyBBJqAAwyYQpKqiIAjDGx5iYiAEwp+DZ2oQTwaMBQVABmTKzkAFwMaAB06GhB7CKNDRXBdWkA3DoA9H1IAPIA0jXMYDAAlizIEGRk5DqFxVQARm7NZDUBXZuV7iLlTQe9ectFJUjHWzVtjVtq9x3kXUYmIFC17SfBvVwDYZjJATaazQgLJZcFZXDaBA47Nx7eHBNQ3U62HA4KbACjzUrOEEHL75fDo4J3DA-Lb-Ahw27+JEBGmYnAwqiFYJQcbEtSkpCA0Y8yYzOaQsg6ekI74PA5vYzZAWDIVEkXg+aLCVccnVRlkZFbeW8GRWbpAA)

It seems like the enums are being merged if their names match.

**Comment by vemoo** [comment]
I think I have a fix for this: https://github.com/vemoo/TypeScript/commit/69d9954d5df923a8c5a5d0466c243981d5cbebe2

As is it makes some tests fail:

- the error at https://github.com/vemoo/TypeScript/blob/69d9954d5df923a8c5a5d0466c243981d5cbebe2/tests/cases/compiler/enumAssignmentCompat3.ts#L70 changes because it bails earlier when it sees that the value of `First.E.c` (2) doesn't match `Cd.E.c` (0).
- before https://github.com/vemoo/TypeScript/blob/69d9954d5df923a8c5a5d0466c243981d5cbebe2/tests/cases/compiler/enumLiteralAssignableToEnumInsideUnion.ts#L28 didn't fail, but now it fails because the values are not the same. I think this change is correct.

I'm not sure what the next steps should be @RyanCavanaugh, update the baselines? do not bail early when checking values? maybe create a new diagnostic?

**Comment by aimed** [comment]
Same issue here with `typescript@3.7.5`

**Comment by kiprobinson** [comment]
Possibly related example, if a function parameter's type is an enum, TS will allow any number:

```
enum myenum {
    hello = 1,
    world = 2,
}

const myfunc = (num:myenum):void => console.log(`num=${num}`);

myfunc(1);            // num=1 (expected)
myfunc(myenum.hello); // num=1 (expected)

// I'm expecting next line to be a TS compile error, but it is not
myfunc(7); // num=7
```

**Comment by aztack** [comment]
It looks like `number enum` is treated as `number` while `string enum` treated as some kind of `unique type`
```ts
enum D {Zero = 0}
// no error. as if D treated as `number`
const d1: D = NaN; 
const d2: D = Infinity;
const d3: D = Math.PI;
const d4: D = 0.1
function f(): number { return Math.random()}
const d5: D = f();

enum S {First = 'first'};
const s1: S = S.First;
const s2: S = 'first'; // error as expected
const s3: S = 'foo'; // error as expected
```

**Comment by giopetris** [comment]
The issue continues to exists on Typescript 5.0 😓

**Comment by HolgerJeromin** [comment]
I have a similar problem:

```ts
namespace space1 {
    export enum Position {
        Left = 0,
        Right = 1,
    }
    export function addPosition(position: Position) {}
}
namespace space2 {
    export enum AxisPosition {
        Left = 0,
        Right = 1,
    }
    export enum Position {
        Left = 0,
        Right = 1,
    }
}
const space2axisFoo = space2.AxisPosition.Left
const space2posFoo = space2.Position.Left

space1.addPosition(space2axisFoo); // Argument of type 'AxisPosition.Left' is not assignable to parameter of type 'Position'.
space1.addPosition(space2posFoo); // But this works
```
A different name is not compatible, the same name (in different scope) is.

[playground](https://www.typescriptlang.org/play?ts=5.2.2#code/HYQwtgpgzgDiDGEAEsEQIxIN4Cgn6QgA8YB7AJwBdDgBXMJABVKgEtLXThs8C+AZCADNqAXiQAGADS8++AEqsA5gAsxSdDL4BfWYRIVqQ2sHgcuSEABMrzNueAAKMvc7AAXExbs3ASmy6uqCQqIgocIgATDx8xGRUNPRIAIJErFB2Pha4cgSCIkji0np8iqrqmnq6sQYJEHQMmQ4xuUj56sWtSGVqhRpaBIE48FxQ1KEQkSBpUABipKR9E5EAdKnpTW4r7cOj4xGTLvOL4ssrm1zbwpQ4OBPoK9a23g6Oy9Ppx74A3HcHD08Lk5lkcFj8cEA)

**Comment by Vict0rd** [comment]
I have the similar issue
```ts
namespace UpperCase {
    export enum Enum {
        A = 'A',
        B = 'B'
    }
}

namespace LowerCase {
    export enum Enum {
        A = 'a',
        B = 'b'
    }
}

const upperCaseEnumValue: UpperCase.Enum = getUpperCaseEnumValue()
const lowerCaseEnumValue: LowerCase.Enum = upperCaseEnumValue; // Why is there no error?

// error: Type 'Enum.B' is not assignable to type 'Enum'. Did you mean 'Enum.B'?
// const upperCaseEnumValue2: UpperCase.Enum = UpperCase.Enum.B
// const lowerCaseEnumValue2: LowerCase.Enum = upperCaseEnumValue2; 

function getUpperCaseEnumValue(): UpperCase.Enum {
    return UpperCase.Enum.B
}

console.log(lowerCaseEnumValue) // prints "B"
```

https://www.typescriptlang.org/play?ts=5.3.2#code/HYQwtgpgzgDiDGEAEBVGMICcDCIrIG8BYAKCXKQgA8YB7TAF0uAFcwkBRV94si-gIJIAvEgDkAsQBpS-fgCER4+WNkUAvqU0lSoSLATIAMrQDuWXPiS9+1Oo2ZtO3a2rlIhosSGlu5irwAjVT5ybW1SeFpgKCYWdAs8CC42ADUQABsWCAAuVAScJIA6FPZRAHMIBjQMQvxS9KyIAAoASkjo2KQMs0T67kbsvJNzOogSl1F42stkgczsgG4kAHoVpAB1AAsATyQASygkBi2sZGBaSkxMegB+UlI1q5vMPIAVHYxxUqKVA6OLkw8FB9uVQIEMsgGJcGJ9kGJSmIikgACL7AAmSB2tBYSEgIGA324vzE9xITyiMTiBVmDQWEAATHkan1xqUlCyxhM2L9HutKV0eqNafMmkykCNWdyykhpqy6WLlg8SAAzFjAeAMfbRJCVao0pIK7JtZkG-DS1yhJCYKosTCEzmzaW8kgREgC2iQoo9crNIXy0XZVqrdYwTD7YAMI4AInk0dIQA

**Comment by Robert-Schirmer** [comment]
Should this be closed? The code example provided in the bug description is now fixed in v5.4.5 of typescript and errors as expected, [playground link](https://www.typescriptlang.org/play/?ts=5.4.5#code/HYQwtgpgzgDiDGEAEAxA9mpBvAUE-SEAHjGgE4AuhwArmEgMJoA252eBnASgKIAiSALxIARLz4iANB074A4rx4A5IaIU9lImUgC+OPTlCRYCZACEQZdp2KlK1OoxZtcsguNVj+U7ZwD0fkgAbiDMNMgAJmjQwADkVBBBEMBIYCAU8AAWSABGEBQA7hDJSBQAnjDQvgTqyp5mADIAglxyABIAhF0+bviNAKo89Q2DWpx6BhEQ8MyWyPBowFBUAGYYTKxkAFyoGAB0G+QA3DgBhCTTFBARSAtgMACWzBBWL2RsmS-IefAgNFDIACSsXoICgUAeAHNgA9gJCkItkMlHOVKqVMCBgGgKJ8yDgFksqDlLIdtkgLGQDs4rMI1mhSUcgA)

**Comment by RyanCavanaugh** [comment]
Yep. Thanks!

**Comment by HolgerJeromin** [comment]
> The code example provided in the bug description is now fixed in v5.4.5 of typescript and errors as expected

My issue "different name is not compatible, the same name (in different scope) is." is still open in nightly:

https://github.com/microsoft/TypeScript/issues/29879#issuecomment-1735247024
